File: typed_async/src/ast.rs
#[derive(Debug, Clone)]
pub(crate) enum Type {
    Number,
    #[allow(dead_code)]
    Arrow(Box<Type>, Box<Type>),
}

pub(crate) type Identifier = String;

#[derive(Debug, Clone)]
pub(crate) enum Stmt {
    Let(Identifier, Type, Box<Expr>),
    Fun(Identifier, Box<Function>),
}

#[derive(Debug, Clone)]
pub(crate) struct Function {
    pub params: Params,
    pub return_ty: Type,
    pub body: Expr,
}

pub(crate) type Params = Vec<Param>;

#[derive(Debug, Clone)]
pub(crate) struct Param {
    pub name: Identifier,
    pub ty: Type,
}

#[derive(Debug, Clone)]
pub(crate) enum Expr {
    Factor(Box<Factor>),
    BinaryOp {
        op: TermBinaryOp,
        lhs: Box<Expr>,
        rhs: Box<Expr>,
    },
}

#[derive(Debug, Clone, Copy)]
pub(crate) enum TermBinaryOp {
    Add,
    #[allow(dead_code)]
    Sub,
}

#[derive(Debug, Clone)]
pub(crate) enum Factor {
    Literal(f64),
    Variable(Identifier),
    #[allow(dead_code)]
    Group(Box<Expr>),
    #[allow(dead_code)]
    BinaryOp {
        op: FactorBinaryOp,
        lhs: Box<Factor>,
        rhs: Box<Factor>,
    },
    Call(Box<Factor>, Vec<Expr>),
}

#[derive(Debug, Clone, Copy)]
pub(crate) enum FactorBinaryOp {
    #[allow(dead_code)]
    Mul,
    #[allow(dead_code)]
    Div,
}

File: ocaml/calc/lib/ast.ml
type typ = Number | Arrow of typ list * typ [@@deriving show]

type identifier = string [@@deriving show]

type stmt = Let of identifier * typ * expr | Fun of identifier * func

and func = {params: params; return_ty: typ; body: expr} [@@deriving show]

and params = param list

and param = {name: identifier; ty: typ} [@@deriving show]

and expr =
  | Factor of factor
  | BinaryOp of {op: term_binary_op; lhs: expr; rhs: expr}

and term_binary_op = Add | Sub

and factor =
  | Literal of float
  | Variable of identifier
  | Group of expr
  | BinaryOp of {op: factor_binary_op; lhs: factor; rhs: factor}
  | Call of {func: factor; args: expr list}

and factor_binary_op = Mul | Div
